src/NATS.Client.Core/NatsConnection.RequestSub.cs

```diff
@@ -1,31 +1,167 @@
 using System.Buffers;
+using System.Runtime.CompilerServices;
 using NATS.Client.Core.Internal;
 
 namespace NATS.Client.Core;
 
 public partial class NatsConnection
 {
-    internal async ValueTask<NatsSub> RequestSubAsync(
+    public async ValueTask<NatsMsg<TReply?>?> RequestAsync<TRequest, TReply>(
+        string subject,
+        TRequest? data,
+        NatsPubOpts? requestOpts = default,
+        NatsSubOpts? replyOpts = default,
+        CancellationToken cancellationToken = default)
+    {
+        var opts = SetReplyOptsDefaults(replyOpts);
+
+        await using var sub = await RequestSubAsync<TRequest, TReply>(subject, data, requestOpts, opts, cancellationToken)
+            .ConfigureAwait(false);
+
+        if (await sub.Msgs.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (sub.Msgs.TryRead(out var msg))
+            {
+                return msg;
+            }
+        }
+
+        if (sub is { EndReason: NatsSubEndReason.Exception, Exception: not null })
+        {
+            throw sub.Exception;
+        }
+
+        return null;
+    }
+
+    public ValueTask<NatsMsg<TReply?>?> RequestAsync<TRequest, TReply>(
+        in NatsMsg<TRequest> msg,
+        NatsSubOpts? replyOpts = default,
+        CancellationToken cancellationToken = default) =>
+        RequestAsync<TRequest, TReply>(
+            msg.Subject,
+            msg.Data,
+            requestOpts: new NatsPubOpts { Headers = msg.Headers, },
+            replyOpts,
+            cancellationToken);
+
+    public async ValueTask<NatsMsg?> RequestAsync(
         string subject,
         ReadOnlySequence<byte> payload = default,
         NatsPubOpts? requestOpts = default,
         NatsSubOpts? replyOpts = default,
         CancellationToken cancellationToken = default)
     {
-        var replyTo = $"{InboxPrefix}{Guid.NewGuid():n}";
-        var sub = await SubAsync(replyTo, replyOpts, NatsSubBuilder.Default, cancellationToken).ConfigureAwait(false);
-        await PubAsync(subject, replyTo, payload, requestOpts?.Headers, cancellationToken).ConfigureAwait(false);
-        return sub;
+        var opts = SetReplyOptsDefaults(replyOpts);
+
+        await using var sub = await RequestSubAsync(subject, payload, requestOpts, opts, cancellationToken).ConfigureAwait(false);
+
+        if (await sub.Msgs.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (sub.Msgs.TryRead(out var msg))
+            {
+                return msg;
+            }
+        }
+
+        if (sub is { EndReason: NatsSubEndReason.Exception, Exception: not null })
+        {
+            throw sub.Exception;
+        }
+
+        return null;
     }
 
+    public ValueTask<NatsMsg?> RequestAsync(
```

Comment:
This and a few other request methods that only delegate and do not access non-public members could be kept as extension methods to keep interface size in check. Only though about this now :/